GH-41256: [Format][Docs] Add a canonical extension type specification for JSON (#41257)
apache/arrow: docs/source/format/CanonicalExtensions.rst

  3) Its serialization *must* be described in the proposal and should
     not require unduly implementation work or unusual software dependencies

>      (for example, a trivial custom text format or YAML would be acceptable).

Author: @ianmcook is giving another example ok?

Reviewer: This sentence is **not** an example of a canonical extension type. It describes how canonical extension types _may_ be serialized as metadata.

So this should remain JSON. We would definitely not allow YAML as the serialization format for canonical extension types.

Reviewer: @ianmcook 's suggestion of "a JSON-based format" is correct though.

Author: Got it. Switched to @ianmcook's suggestion.

>      (for example, a trivial custom text format or YAML would be acceptable).

Author: ```suggestion
     (for example, a trivial custom text format or a JSON-based format would be acceptable).
```

   Values inside each **data** tensor element are stored in row-major/C-contiguous
   order according to the corresponding **shape**.

.. _json_extension:

JSON
====

* Extension name: `arrow.json`.

* The storage type of this extension is ``StringArray`` or
  ``StringViewArray``.

* Extension type parameters:

  This type does not have any parameters.

* Description of the serialization:

>   Metadata is an empty string.

Reviewer 2: Perhaps we could have a mechanism to encode JSON schema or similar? :thinking:

Reviewer 3: That would be interesting, though I'm not sure about adding that until we have someone who will definitely implement using it.

We could phrase this in a way that metadata is currently optional. Maybe:

```suggestion
  Metadata is either an empty string or a JSON string with an empty object.
  In the future, additional fields may be added, but they are not required
  to interpret the array.
```

Author: Year, that makes sense as we are promising [backwards compatibility](https://arrow.apache.org/docs/format/CanonicalExtensions.html#making-modifications):

> Making Modifications
> Like standard Arrow data types, canonical extension types should be considered stable once standardized. Modifying a canonical extension type (for example to expand the set of parameters) should be an exceptional event, follow the same rules as laid out above, and provide backwards compatibility guarantees.

   Values inside each **data** tensor element are stored in row-major/C-contiguous
   order according to the corresponding **shape**.

.. _json_extension:

JSON
====

* Extension name: `arrow.json`.

* The storage type of this extension is ``StringArray`` or

>   ``LargeStringArray``or ``StringViewArray``.

Reviewer: I think it's ok, but the Large variants can perhaps be omitted since we're unlikely to store a JSON value more than 2 GB large.

Similarly, for `arrow.variable_shape_tensor` we're mandating a List type, not LargeList.

Reviewer: Perhaps mention that only UTF-8 encodings of JSON are supported.

Author: Removing and waiting for others to chime in.

Reviewer 3: > the Large variants can perhaps be omitted

Some Arrow-native libraries (such as Polars, before it switched to use `StringView`) might work exclusively work large variants. I'm inclined to include them.

Author: I don't see a downside allowing for `LargeStringArray` even though it doesn't make much sense.

Reviewer: Fair enough!

* Extension type parameters:

  This type does not have any parameters.

* Description of the serialization:

  Metadata is an empty string.

>   - Example: ``""``

Reviewer: I don't think the example is useful ;-)

Author: It's great. It covers all cases. Removing.

   Values inside each **data** tensor element are stored in row-major/C-contiguous
   order according to the corresponding **shape**.

.. _json_extension:

JSON
====

* Extension name: `arrow.json`.

> * The storage type of this extension is ``StringArray`` or
>   ``StringViewArray``.

Author: @pitrou 
```suggestion
* The storage type of this extension is ``StringArray`` or
  ``StringViewArray``. Only UTF-8 encoded JSON is supported.
```

Reviewer 4: Should one of those say "LargeString"?

Author: Indeed. Thanks! :D

   Values inside each **data** tensor element are stored in row-major/C-contiguous
   order according to the corresponding **shape**.

.. _json_extension:

JSON
====

> * Extension name: `arrow.json`.

Reviewer: ```suggestion
* Extension name: ``arrow.json``.
```

* The storage type of this extension is ``StringArray`` or

>   or ``LargeStringArray`` or ``StringViewArray``.

Reviewer 5: One further nit: "StringArray" is not really generic terminology that we use in the format specs (it's the name of the array/vector in Arrow C++ and pyarrow, but not necessarily in other implementations). So I think you can leave out the `Array` part in each of those three types. 

(strictly speaking the spec actually uses Utf8 / LargeUtf8 / Utf8View, but I think in many places we do use "String" instead)

Reviewer: Let's use the same wording as in the spec, then?

Author: @jorisvandenbossche like so?
```suggestion
* The storage type of this extension is ``String`` or
  or ``LargeString`` or ``StringView``.
```

>   Only UTF-8 encoded JSON is supported.

Reviewer 6: I think the one thing to consider here as possible future extensions (this might have already been discussed on ML a while ago), is non-standard encodings using Binary types instead of String.

In the short term we should probably recommend that implementations should likely reject Binary annotated JSON fields

Reviewer 7: I would suggest that binary encodings of JSON should be advertized as a different extension, e.g. "arrow.bson"

Reviewer 6: Sorry to clarify I was not talking about BSON, but rather UTF-16 or UTF-32 encodings for JSON (which one can debate if it is worth supporting based on RFC working about requiring UTF-8).

Reviewer 8: > By this you mean serializing binary data into UTF-8 string and including it as a String? What would be a downside to allowing that?

I think the argument was that we might want to support `BinaryArray` as a storage type since a JSON document could be encoded with a non-utf8 encoding and thus should not be stored in a `StringArray`.  However, I agree we probably don't want to worry about this since the RFC is pretty clear that JSON must be UTF-8 and users can always make up their own extension type if they need to.

* Extension type parameters:

  This type does not have any parameters.

* Description of the serialization:

>   Metadata is either an empty string or a JSON string with an empty object.

Reviewer 5: In addition to being an empty string, is it also allowed to be just absent?

Reviewer: For now I think we should think of the extension serialization metadata key  as mandatory.

  In the future, additional fields may be added, but they are not required

>   to interpret the array.

Reviewer 5: Is there a specific reason we want to add this here for JSON, while we don't do this for uuid? Do we expect there to be metadata in the future?

(I would also argue that adding _optional_ metadata in the future should be regarded as backwards compatible, even if we don't specify that here now)

Reviewer 7: An interesting metadata that could be put is ``{"schema": { ... insert here JSON schema .... }}`` to express contraints on the payload

Author: @tustvold suggested metadata could hold JSON schema which would not be the case with UUID which could justify the difference between the types.
I do agree with your reasoning that adding optional metadata later wouldn't be backwards incompatible, but perhaps we have users that need it now. @progger-dev any thoughts from you or someone on the bigquery team?

Reviewer 3: > I would also argue that adding optional metadata in the future should be regarded as backwards compatible, even if we don't specify that here now

I agree, but in practice what is backwards compatible depends on what implementations do. So I think it's worth being extra clear here that it should not be an error if there is unexpected metadata.

Reviewer 9: I had 3 things in my mind for what I would like to be able to use the metadata field for.

1. Parsing options that describe what features were used to generate the JSON. E.g. trailing commas allowed, unquoted field names, etc.
2. Schemas
3. Specialized metadata used internally (e.g. BigQuery might want to store what fields were columnarized, so we could reinstantiate it if the user exports data and then reimports it back).

None of these were requirements, but just some potential things that we might have used the metadata field for. I'm no longer at Google, so I'll tag @emkornfield for more recent thoughts from the BQ team.

Reviewer 7: > 1. Parsing options that describe what features were used to generate the JSON. E.g. trailing commas allowed, unquoted field names, etc

ouch, I wouldn't call JSON with trailing commas, or unquoted field names JSON. Perhaps it would be wise to specify that JSON here is the one of https://www.json.org/json-en.html , that is ECMA-404 , otherwise that would defeat the purpose of interoperability between implementations

Reviewer 9: I totally agree. I don't think we should canonicalize these non-standard extensions. I don't think producers should generate non-standard JSON. But I think consumers should be allowed to parse non-standard JSON. This is consistent with the spec in [RFC-8259](https://datatracker.ietf.org/doc/html/rfc8259#section-9).

This also begs the question of what we do when JSON parsing fails.

Reviewer 10: Should we explicitly state in `CanonicalExtensions.rst` that this canonical extension type supports JSON _as defined in RFC-8259 (https://datatracker.ietf.org/doc/html/rfc8259)_?

Reviewer 5: > 3\. Specialized metadata used internally (e.g. BigQuery might want to store what fields were columnarized, so we could reinstantiate it if the user exports data and then reimports it back).

Note that for such internal metadata, you are always free that any other key to the field metadata of the column, and you don't necessarily have to add that to the value of the `ARROW:extension:metadata` key

Reviewer 8: I think I agree with Ian that stricter is better.  It's fine for different components within a system to agree they want to support a wider set of values.  However, once data leaves a solution, it should assume that most components will only be processing the to-spec JSON.  This document is primarily about interchange and so should be concerned with the latter case (data leaving a solution to be consumed by some unknown component).

Author: Added rfc8259 requirement (https://github.com/apache/arrow/pull/41257/commits/594945010e3b7d393b411aad971743ffcdbdbc8e).

To the original @jorisvandenbossche's question: it seems to me we're more in favour of adding optional metadata to JSON type than not so I'm leaving that language as is. Please correct me if I'm wrong.

Author: @progger-dev is 1. or 2. like what you have in mind?
1 . `Non-conformant or non-parsable JSON values should be considered invalid and halt producing or consuming process.`
2. `Non-conformant or non-parsable JSON values should be considered as null values by producers and parsers.`

Reviewer 9: I was thinking about treating it the same as if you didn't support the extension type itself and just dealt with the underlying type and acts as a pass through.

But on second thought, this is not something for the specification as its likely going to be highly implementation dependent. Maybe its more of a recommendation?

Reviewer 5: > I think we should also add language that specifies what happens when you receive non-conformant JSON or see any kind of parse errors. I think it's important for components that sit in between producers and consumers (and thus act as both).

Is it needed to specify anything about that? To me that seems a decision that an application that will actually use the JSON data will need or want to decide for themselves how to treat invalid input, while in many cases an Arrow implementation receiving such data might just be passing it along, and will typically not do any validation at all.

   Values inside each **data** tensor element are stored in row-major/C-contiguous
   order according to the corresponding **shape**.

.. _text_data_format_extension:

Text data format
=================

* Extension name: `arrow.text_data_format`.

* The storage type of the extension is ``StringArray`` or

>   ``LargeStringArray``or ``StringViewArray``.

Author: @pitrou any thougts on storage type?
